### DISCUSSION TITLE
Feyncalc computations using parallel computing (multicore computing)?

### DISCUSSION THREAD

**Discussion by armandlcz** [FeynCalc/feyncalc#258] [Q&A]
Hello FeynCalc team!

First of all, I want to thank you for this fantastic project; the world would not be the same without FeynCalc. Well, let me explain my question. I'm doing very long computations (they're dimensional regularizations of diagrams with a lot of D-dimensional metric and D-dimensional momentum contractions like Contract[ MTD[\[Alpha,\[Beta] MTD[\[Beta],\[Gamma] FVD[p,\[Delta]... ] involved). For the first diagram I computed, it took almost 6 days working full time to give the result (my laptop is corei5 10th gen, 4 cores, 32GB RAM). Looking into Mathematica documentation, I found that it's possible to do computations faster using parallel computing, i.e. using more than one kernel/core to do one single computation. This can be done with the Parallelize[] function. Unfortunately, when I try to parallelize the contraction expression (for example, Parallelize[ Contract[ MTD[ \[Alpha] \[Beta] ] MTD[ \[Alpha] \[Beta] ] ] ] ), I get the message "(expr) cannot be parallelized; proceeding with sequential evaluation". I wonder if there is some way to use parallel computing in FeynCalc (to take advantage of all the CPU's cores available) or if any other way exists to make computations faster. In fact, I have access to a 16-core PC at my university, but I think this PC would not be helpful as long as parallel computing is not possible, right? Any suggestions?

Well, I hope you can help me. Thank you again for this great project!

**Comment by vsht** [comment]
There is almost no built-in parallelization in FeynCalc, so using ParallelMap or just evaluating every diagram separately is the best you can do. 

Having said that, 6 days for a diagram is such a long runtime, that I would seriuously consider switching to FORM at this stage. Obviously, the complexity of your calculation is too high for Mathematica to handle.
